src/scrolling.c

```diff
@@ -0,0 +1,97 @@
+/*
+ * scrolling.c: miscellaneous functions for scrollable widgets
+ * Gtkdialog - A small utility for fast and easy GUI building.
+ * Copyright (C) 2021       step https://github.com/step-
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+/* Includes */
+#define _GNU_SOURCE
+#include <gtk/gtk.h>
+#include "config.h"
+#include "gtkdialog.h"
+#include "attributes.h"
+#include "automaton.h"
+#include "widgets.h"
+#include "signals.h"
+#include "tag_attributes.h"
+#include "scrolling.h"
+
+/* Defines */
+//#define DEBUG_CONTENT
+//#define DEBUG_TRANSITS
+
+/* Local function prototypes, located at file bottom */
+static void scroll_value_changed (GtkAdjustment *adjustment, gpointer user_data);
+static void scroll_bottom_gravity (GtkWidget *scrolled_window,
+	GdkRectangle *allocation,
+	gpointer      user_data);
+
+/* Notes:
+ * step - scroll to bottom inspired by https://discourse.gnome.org/t/1337/2 */
+
+/***********************************************************************
+ * Scroll to Bottom
+ ***********************************************************************/
+
+void setup_scroll_to_bottom(variable *var)
+{
+	/* var->Widget being a scrollable widget can be scrolled to bottom */
+	GtkAdjustment *adjustment;
+	GtkWidget     *w;
+	gchar         *value;
+	double *from_bottom;
+
+	/* If the custom attribute "scroll-to-bottom" is true
+	 * then arrange for the scrollbar to scroll to bottom */
+	if (var->widget_tag_attr &&
+		(value = get_tag_attribute(var->widget_tag_attr, "scroll-to-bottom")) &&
+		((strcasecmp(value, "true") == 0) ||
+		(strcasecmp(value, "yes") == 0) || (atoi(value) == 1)) &&
+		(w = gtk_widget_get_ancestor(var->Widget, GTK_TYPE_SCROLLED_WINDOW)) &&
+		(from_bottom = g_slice_new0(double))) /* from_bottom is never freed */
```

Comment:
Why is it OK not to free it?

Reply:
It can be frowned upon but consider that this function `setup_scroll_to_bottom` is called only once when the widget is set up; so one could view from_bottom's memory (a double) as part of the widget's memory allocation. This code isn't called repeatedly for a given widget, so there shouldn't be a memory leak.